src/platform/backends/lxd/lxd_vm_image_vault.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "lxd_vm_image_vault.h"
+
+#include <multipass/exceptions/aborted_download_exception.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/rpc/multipass.grpc.pb.h>
+#include <multipass/url_downloader.h>
+#include <multipass/vm_image.h>
+#include <multipass/vm_image_host.h>
+
+#include <QRegularExpression>
+
+#include <chrono>
+#include <thread>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+using namespace std::literals::chrono_literals;
+
+namespace
+{
+constexpr auto category = "image vault";
+
+auto parse_percent_as_int(const QString& progress_string)
+{
+    QRegularExpression re{"\\s(\\d{1,3})%"};
+
+    QRegularExpressionMatch match = re.match(progress_string);
+
+    if (match.hasMatch())
+    {
+        return match.captured(1).toInt();
+    }
+
+    return -1;
+}
+} // namespace
+
+mp::LXDVMImageVault::LXDVMImageVault(std::vector<VMImageHost*> image_hosts, const QUrl& base_url)
+    : image_hosts{image_hosts}, base_url{base_url}, manager{std::make_unique<mp::NetworkAccessManager>()}
+{
+    for (const auto& image_host : image_hosts)
+    {
+        for (const auto& remote : image_host->supported_remotes())
+        {
+            remote_image_host_map[remote] = image_host;
+        }
+    }
+}
```

Comment:
Should we filter out those without a stream URL here?

Reply:
> I'm pretty sure I already did that for cloud images. I know I have not accounted for our custom repository yet. Is this what you are seeing?

This is probably what could achieve this?

Reply:
Yep, that is what I'll do for now until we can support the custom repository and http/file based launches.

Reply:
After looking at this, the image host itself has no knowledge if `stream_location` is empty or now.  It would have to iterate through each image that the image host knows about to see if `stream_location` is known.  I can do this, but it's just going to be a hack for now.

Reply:
In order to match what other platforms do, I implemented this in `mp::platform::is_remote_supported()` and that takes care of the remotes part.  I then added logic to throw in `fetch_image()` for URL based launches.